Devirtualize numel()

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27294 Devirtualize numel()**

Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17763486](https://our.internmc.facebook.com/intern/diff/D17763486)

ezyang committed:
Devirtualize numel()

Fixes #27291

I think this deletes at::numel function too, making it technically BC-breaking,
but I don't think we intended to generate that function in the first place.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
bring back torch.numel on "Devirtualize numel()"


Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize numel()"


Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ailzhang this one breaks xla but it should be a pretty simple fix; just delete the numel() implementation. (This is under the assumption that your TensorImpl::numel implementation does the right thing. Which it should.)

ailzhang commented:
Thanks @ezyang , I'm starting a local run to verify the change. :D Will report back once it's finished.

ezyang committed:
Update on "Devirtualize numel()"


Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ailzhang commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -369,14 +369,15 @@ def declkey(decl):
         return [sorted(g, key=declkey) for g in grouped_decls]
 
     # We need to add methods implemented manually in TensorImpl
+    # TODO: This seems to claim sizes() returns an int64_t.  Really?
```
our `Tensor::sizes()` does return a int64_t iirc. It's number of dimensions. And yes I agree it's confusing :P

ailzhang reviewed (approved):
Fix is in https://github.com/pytorch/xla/pull/1136

ezyang committed:
Update on "Devirtualize numel()"


Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize numel()"


Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17763486](https://our.internmc.facebook.com/intern/diff/D17763486)

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize numel()"


Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17763486](https://our.internmc.facebook.com/intern/diff/D17763486)

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize numel()"


Fixes #27291

I'm a little annoyed that I have to reintroduce manual binding code.  But it's
probably not a good idea to teach the codegen how to do fastpath functions
(is it?)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17763486](https://our.internmc.facebook.com/intern/diff/D17763486)

[ghstack-poisoned]
